File: android/src/main/java/io/appmetrica/analytics/expo/LocationInfo.kt
package io.appmetrica.analytics.expo

import android.location.Location
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class LocationInfo: Record {
    @Field
    val latitude: Double = 0.0
    @Field
    val longitude: Double = 0.0
    @Field
    val altitude: Double? = null
    @Field
    val accuracy: Float? = null
    @Field
    val course: Float? = null
    @Field
    val speed: Float? = null
    @Field
    val timestamp: Long? = null

    fun toLocation(): Location {
        val location = Location("Custom")
        location.latitude = latitude
        location.longitude = longitude
        altitude?.let { location.longitude = it }
        accuracy?.let { location.accuracy = it }
        course?.let { location.bearing = it }
        speed?.let { location.speed = it }
        timestamp?.let { location.time = it }
        return location
    }
}

File: ios/LocationInfo.swift

import ExpoModulesCore

struct LocationInfo: Record {
    @Field
    var latitude: Double = 0.0
    @Field
    var longitude: Double = 0.0
    @Field
    var altitude: Double? = nil
    @Field
    var accuracy: Float? = nil
    @Field
    var verticalAccuracy: Float? = nil
    @Field
    var course: Float? = nil
    @Field
    var speed: Float? = nil
    @Field
    var timestamp: Double? = nil

    func toLocation() -> CLLocation {
        let date = Date(timeIntervalSince1970: TimeInterval(timestamp!))
        let coordinate = CLLocationCoordinate2D(latitude: CLLocationDegrees(latitude), longitude: CLLocationDegrees(longitude))
        let location = CLLocation(coordinate: coordinate, altitude: CLLocationDistance(altitude!), horizontalAccuracy: CLLocationAccuracy(accuracy!), verticalAccuracy: CLLocationAccuracy(verticalAccuracy!), timestamp: date) 
        return location
    }
}